draft-irtf-cfrg-bbs-signatures.md

```diff
@@ -285,53 +285,48 @@ Those rules will be used explicitly on every operation. See also [Serialize](#se
-
-KeyGen uses an HKDF [@!RFC5869] instantiated with the hash function hash.
+This operation generates a secret key (SK) deterministically from a secret octet string (key\_material). This operation is the RECOMMENDED way of generating a secret key, but its use is not required for compatibility, and implementations MAY use a different KeyGen procedure. For security, such an alternative MUST output a secret key that is statistically close to uniformly random in the range from 1 to r-1. An example of an HKDF-based alternative is the KeyGen operation, defined in Section 2.3 of [@I-D.irtf-cfrg-bls-signature] (with an appropriate, BBS specific, salt value, like "BBS_SIG_KEYGEN_SALT_").
 
 For security, IKM MUST be infeasible to guess, e.g. generated by a trusted source of randomness.
 
-IKM MUST be at least 32 bytes long, but it MAY be longer.
+key\_material MUST be at least 32 bytes long, but it MAY be longer.
 
-Because KeyGen is deterministic, implementations MAY choose either to store the resulting SK or to store IKM and call KeyGen to derive SK when necessary.
+Because KeyGen is deterministic, implementations MAY choose either to store the resulting SK or to store key\_material and call KeyGen to derive SK when necessary.
 
-KeyGen takes an optional parameter, key\_info. This parameter MAY be used to derive multiple independent keys from the same IKM.  By default, key\_info is the empty string.
+KeyGen takes an optional input, key\_info. This parameter MAY be used to derive multiple independent keys from the same key material.
 
 ```
-SK = KeyGen(IKM, key_info)
+SK = KeyGen(key_material, key_info)
 
 Inputs:
 
-- IKM (REQUIRED), a secret octet string. See requirements above.
-- key_info (OPTIONAL), an octet string. if this is not supplied, it
-                       MUST default to an empty string.
+- key_material (REQUIRED), a secret octet string. See requirements
+                           above.
+- key_info (OPTIONAL), an octet string. Defaults to an empty string if
+                       not supplied.
 
 Definitions:
 
-- HKDF-Extract is as defined in [@!RFC5869], instantiated with hash function hash.
-- HKDF-Expand is as defined in [@!RFC5869], instantiated with hash function hash.
-- I2OSP and OS2IP are as defined in [@!RFC8017], Section 4.
-- L is the integer given by ceil((3 * ceil(log2(r))) / 16).
-- INITSALT is the ASCII string "BBS-SIG-KEYGEN-SALT-".
+- key_dst is the ASCII string "BBS-SIG-KEYGEN-SALT-".
 
 Outputs:
 
 - SK, a uniformly random integer such that 0 < SK < r.
 
 Procedure:
 
-1. salt = INITSALT
-2. SK = 0
-3. while SK == 0:
-4.     salt = hash(salt)
-5.     PRK = HKDF-Extract(salt, IKM || I2OSP(0, 1))
-6.     OKM = HKDF-Expand(PRK, key_info || I2OSP(L, 2), L)
-7.     SK = OS2IP(OKM) mod r
-8. return SK
+1.  if length(key_material) < 32, return INVALID
+2.  if length(key_info) > 65535, return INVALID
+3.  derive_input = key_material || I2OSP(length(key_info), 2) || key_info
+4.  counter = 0
+5.  SK = 0
+6.  while SK == 0:
```

Comment:
Just a note: `hash_to_scalar` currently guarantees that `SK !- 0`. IMO of course it would be better to update `hash_to_scalar` than this operation.